multiraft/multiraft.go

```diff
@@ -696,30 +703,49 @@ func (s *state) handleWriteResponse(response *writeResponse, readyGroups map[uin
 				if err != nil {
 					log.Fatalf("invalid ConfChange data: %s", err)
 				}
+				var payload []byte
 				if len(cc.Context) > 0 {
-					commandID, _ = decodeCommand(cc.Context)
-				}
-				log.V(3).Infof("node %v applying configuration change %v", s.nodeID, cc)
-				// TODO(bdarnell): dedupe by keeping a record of recently-applied commandIDs
-				switch cc.Type {
-				case raftpb.ConfChangeAddNode:
-					err = s.addNode(NodeID(cc.NodeID), groupID)
-				case raftpb.ConfChangeRemoveNode:
-					// TODO(bdarnell): support removing nodes; fix double-application of initial entries
-					continue
-				case raftpb.ConfChangeUpdateNode:
-					// Updates don't concern multiraft, they are simply passed through.
-				}
-				if err != nil {
-					log.Errorf("error applying configuration change %v: %s", cc, err)
+					commandID, payload = decodeCommand(cc.Context)
 				}
-				cs := s.multiNode.ApplyConfChange(groupID, cc)
 				s.sendEvent(&EventMembershipChangeCommitted{
 					GroupID:    groupID,
 					CommandID:  commandID,
 					NodeID:     NodeID(cc.NodeID),
 					ChangeType: cc.Type,
-					ConfState:  *cs,
+					Payload:    payload,
+					Callback: func(err error) {
+						s.callbackChan <- func() {
+							if err == nil {
+								log.V(3).Infof("node %v applying configuration change %v", s.nodeID, cc)
+								// TODO(bdarnell): dedupe by keeping a record of recently-applied commandIDs
+								switch cc.Type {
+								case raftpb.ConfChangeAddNode:
+									err = s.addNode(NodeID(cc.NodeID), groupID)
+								case raftpb.ConfChangeRemoveNode:
+									// TODO(bdarnell): support removing nodes; fix double-application of initial entries
+								case raftpb.ConfChangeUpdateNode:
+									// Updates don't concern multiraft, they are simply passed through.
+								}
+								if err != nil {
+									log.Errorf("error applying configuration change %v: %s", cc, err)
+								}
+								s.multiNode.ApplyConfChange(groupID, cc)
+							} else {
+								log.Warningf("aborting configuration change: %s", err)
+								s.multiNode.ApplyConfChange(groupID,
```

Comment:
[From my snippet above](https://github.com/cockroachdb/cockroach/pull/364/files#diff-dfd5fe6017011d7ffff21e9d95d739c3R175) it looks like Raft simply discards the new ConfChange rather than abort the previous one, what am I missing?